mne/preprocessing/bads/bads.py

```diff
@@ -0,0 +1,278 @@
+# Authors: Denis Engemann <denis.engemann@gmail.com>
+#          Marijn van Vliet <w.m.vanvliet@gmail.com>
+# License: BSD (3-clause)
+
+import numpy as np
+from ...utils import verbose
+from ...io.pick import pick_info
+from ...io.pick import pick_types
+from . import faster_ as _faster
+
+
+@verbose
+def find_bad_channels(epochs, picks=None, method='faster', method_params=None,
+                      return_by_metric=False, verbose=None):
+    """Implements the first step of the FASTER algorithm.
+
+    This function attempts to automatically mark bad EEG channels by performing
+    outlier detection. It operated on epoched data, to make sure only relevant
+    data is analyzed.
+
+    Parameters
+    ----------
+    epochs : Instance of Epochs
+        The epochs for which bad channels need to be marked
+    picks : list of int | None
+        Channels to operate on. Defaults to EEG channels.
+    method : {'faster'}
+        The detection algorithm.
+    method_params : dict | None
+        The method parameters in a dict.
+
+        If ``method`` equals 'faster', and ``method_params``is None,
+        defaults to the following parameters. Partial updates are supported.
+        use_metrics : list of str
+            List of metrics to use. Can be any combination of:
+                'variance', 'correlation', 'hurst', 'kurtosis', 'line_noise'
+            Defaults to all of them.
+        thresh : float
+            The threshold value, in standard deviations, to apply. A channel
+            crossing this threshold value is marked as bad. Defaults to 3.
+        max_iter : int
+            The maximum number of iterations performed during outlier detection
+            (defaults to 1, as in the original FASTER paper).
+
+    return_by_metric : bool
+        Whether to return the bad channels as a flat list (False, default) or
+        as a dictionary with the names of the used metrics as keys and the
+        bad channels found by this metric as values. Is ignored if not
+        supported by method.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+        Defaults to self.verbose.
+
+    Returns
+    -------
+    bads : list of str
+        The names of the bad EEG channels.
+    """
+    if picks is None:
+        picks = pick_types(epochs.info, meg=True, eeg=True, exclude=[])
+
+    if method == 'faster':
+        _method_params = dict(max_iter=1, thresh=3, use_metrics=None)
+        if isinstance(method_params, dict):
+            _method_params.update(method_params)
+        method_params = _method_params
+        bads = _faster._find_bad_channels(epochs, picks, **_method_params)
+    else:
+        raise NotImplementedError(
+            'Come back later, for now there is only "FASTER"')
+
+    if return_by_metric:
+        return bads
+    else:
+        return np.unique(np.concatenate(bads.values())).tolist()
```

Comment:
Looks good. Now we need to develop the method we were discussing and plug
it there.
Le 13 juil. 2015 10:56 AM, "Denis A. Engemann" <notifications@github.com> a
écrit :

> In mne/preprocessing/bads/bads.py
> <https://github.com/mne-tools/mne-python/pull/1777#discussion_r34464565>:
>
> > +        picks = pick_types(epochs.info, meg=True, eeg=True, exclude=[])
> > +
> > +    if method == 'faster':
> > +        _method_params = dict(max_iter=1, thresh=3, use_metrics=None)
> > +        if isinstance(method_params, dict):
> > +            _method_params.update(method_params)
> > +        method_params = _method_params
> > +        bads = _faster._find_bad_channels(epochs, picks, **_method_params)
> > +    else:
> > +        raise NotImplementedError(
> > +            'Come back later, for now there is only "FASTER"')
> > +
> > +    if return_by_metric:
> > +        return bads
> > +    else:
> > +        return np.unique(np.concatenate(bads.values())).tolist()
>
> @fraimondo <https://github.com/fraimondo> cc look here, a new API that
> makes us independent of the method.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1777/files#r34464565>.
>